### PR TITLE
Added ShoppingCart, CartItems entity and methods

### PR DESCRIPTION
**Pull request by hanshyn**
Shopping Cart Management:

As a User, I want to add books to my shopping cart, so I can purchase them. I will:
Send a POST request to /api/cart to add a book to the shopping cart.
As a User, I want to view my shopping cart, so I can review its contents before placing an order. I will:
Send a GET request to /api/cart to retrieve my shopping cart.
As a User, I want to update the books quantity in my shopping cart, so I can update my intended purchases. I will:
Send a PUT request to /api/cart/cart-items/{cartItemId} to update the books quantity in the shopping cart.
As a User, I want to remove books from my shopping cart, so I can update my intended purchases. I will:
Send a DELETE request to /api/cart/cart-items/{cartItemId} to remove a book from the shopping cart.

### CODE REVIEW

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+        ShoppingCartDto shoppingCartDto = shoppingCartMapper.toDto(shoppingCart);
+        shoppingCartDto.setCartItems(cartItemsResponseDtos);
+
+        return shoppingCartDto;
+    }
+
+    @Override
+    public CartItemsResponseDto saveBookToShoppingCart(Authentication authentication,
+                                                       CartItemsRequestDto cartItemsRequestDto) {
+        User user = (User) authentication.getPrincipal();
```
Comment (olekskov) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: 1) For me it's better to leave work with authentication on controller level and let service work with entity required
2) Several DB modifications inside method, without proper annotation we may loose consistency of data in DB
```suggestion
    @Override
    @Transactional
    public CartItemsResponseDto saveBookToShoppingCart(User user,
                                                       CartItemsRequestDto cartItemsRequestDto) {
```

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createNewShoppingCart(user));
```
Comment (olekskov) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: Sorry, but I cannot understand logic.
If we try to get ShoppingCart and haven't found it we throw exception. But if we adding item and haven't found cart we create it.
How should get operation work before first insert?

In general: shop is place where we need users in order them to buy something. User without cart has no any value.
For me it's more logical to create cart during registration process and assume that it's exist by default and handle case when cart wasn't found as some data integrity violation

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+        cartItem.setQuantity(cartItemsUpdateRequestDto.getQuantity());
+
+        return cartItemMapper.toDo(cartItemRepository.save(cartItem));
+    }
+
+    @Override
+    public void deleteCartItems(Long id) {
+        cartItemRepository.deleteById(id);
+    }
```
Comment (olekskov) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: On add operation you check user that is authenticated but on update/delete you allow anyone to modify other user's cart. 
E.g. I'm user 1 and have items 1, 2 and 3 in my cart. I'm sending request to delete cart-item with id 4. Then you will delete it for some other user obviously

**Review by olekskov — CHANGES_REQUESTED** [review]

File: src/main/java/hanshyn/onlinebookstore/mapper/CartItemMapper.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import hanshyn.onlinebookstore.config.MapperConfig;
+import hanshyn.onlinebookstore.dto.cart.CartItemsRequestDto;
+import hanshyn.onlinebookstore.dto.cart.CartItemsResponseDto;
+import hanshyn.onlinebookstore.model.CartItem;
+import org.mapstruct.Mapper;
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class)
+public interface CartItemMapper {
+
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/mapper/CartItemMapper.java]:
```suggestion
```

File: src/main/java/hanshyn/onlinebookstore/model/ShoppingCart.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import lombok.EqualsAndHashCode;
+import org.apache.commons.lang3.builder.ToStringExclude;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Data
+@Entity
+@Table(name = "shopping_carts")
+@SQLDelete(sql = "UPDATE shopping_cart SET is_deleted = true WHERE id = ?")
+@Where(clause = "is_deleted = false")
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/model/ShoppingCart.java]:
```suggestion
@SQLDelete(sql = "UPDATE shopping_cart SET is_deleted = TRUE WHERE id = ?")
@Where(clause = "is_deleted = FALSE")
```

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartService.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.security.core.Authentication;
+
+public interface ShoppingCartService {
+    ShoppingCartDto getShoppingCart(Authentication authentication);
+
+    CartItemsResponseDto saveBookToShoppingCart(User user,
+                                                CartItemsRequestDto cartItemsRequestDto);
+
+    CartItemsResponseDto updateQuantity(
+                                        CartItemsUpdateRequestDto cartItemsUpdateRequestDto,
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartService.java]:
```suggestion
    CartItemsResponseDto updateQuantity(CartItemsUpdateRequestDto cartItemsUpdateRequestDto,
```

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartDto getShoppingCart(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId()).orElseThrow(
+                () -> new RuntimeException("Can't found user: user")
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: too generic, use EntityNotFound or create your own UserNotFoundException

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+        return shoppingCartDto;
+    }
+
+    @Override
+    @Transactional
+    public CartItemsResponseDto saveBookToShoppingCart(User user,
+                                                       CartItemsRequestDto cartItemsRequestDto) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new RuntimeException("Can]t found user: " + user));
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java:62]: same

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+    private CartItem getCartItemById(User user, Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new RuntimeException("Can't found user by id: " + user.getId()));
+
+        Set<CartItem> cartItemSet = cartItemRepository.findByShoppingCartId(shoppingCart.getId());
+
+        CartItem cartItem = cartItemSet.stream()
+                .filter(cart -> cart.getId().equals(id))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Can't found cart items by id :" + id));
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: entity not found

File: src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    @Override
+    public void deleteCartItems(Long id) {
+        User user = (User) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
+        CartItem cartItem = getCartItemById(user, id);
+        cartItemRepository.delete(cartItem);
+    }
+
+    private CartItem getCartItemById(User user, Long id) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new RuntimeException("Can't found user by id: " + user.getId()));
```
Comment (okuzan) [inline src/main/java/hanshyn/onlinebookstore/service/cart/ShoppingCartServiceImpl.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

**Review by Vasul-Pankiv — APPROVED** [review]

**Pull request merged**